File: leetcode/0148-sort-list/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head :
            return head
        curr = head
        arr =[]
        while curr:
            arr.append(curr.val)
            curr = curr.next
        arr.sort()
        m = ListNode(0)
        curr = m
        for i in range(len(arr)):
            node = ListNode(arr[i])
            curr.next = node
            curr = curr.next
        return m.next


        

        

File: leetcode/0148-sort-list/solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var sortList = function(head) {
    if(!head || ! head.next){
        return head
    }
    let fast = head
    let slow = head
    let prev = null
    while(fast && fast.next){
        prev = slow
        slow=slow.next
        fast = fast.next.next
    }
    prev.next = null
    let left = sortList(head)
    let right = sortList(slow)
   return merge(left,right)
};
const merge=(l1,l2)=>{
    let dummy = new ListNode(0)
    let tail = dummy
    while(l1&&l2){
        if(l1.val<=l2.val){
            tail.next=l1
            l1=l1.next
        }
        else{
             tail.next=l2
             l2=l2.next
        }
        tail = tail.next
    }
    tail.next=l1||l2
    return dummy.next
}
